### ISSUE TITLE
Improper TEXT decoding of some characters in sub-marshallables can result in runtime exceptions

### ISSUE THREAD

**Issue by devinrsmith** [OpenHFT/Chronicle-Wire#38]
The following code results in:

```
output: [obj: {
  "{"
}
]
Exception in thread "main" net.openhft.chronicle.core.io.IORuntimeException: Unterminated { while reading marshallable {,code='￿', bytes=
	at net.openhft.chronicle.wire.TextWire$TextValueIn.marshallable(TextWire.java:2240)
	at net.openhft.chronicle.wire.ValueIn.marshallable(ValueIn.java:319)
	at com.redcurrent.backend.push.gcm.marshallable.MarshallableObj$Outer.readMarshallable(MarshallableObj.java:68)
	at com.redcurrent.backend.push.gcm.marshallable.MarshallableObj.main(MarshallableObj.java:108)
```

```java
import com.google.common.base.Preconditions;
import net.openhft.chronicle.bytes.Bytes;
import net.openhft.chronicle.core.io.IORuntimeException;
import net.openhft.chronicle.wire.Marshallable;
import net.openhft.chronicle.wire.WireIn;
import net.openhft.chronicle.wire.WireOut;
import net.openhft.chronicle.wire.WireType;
import org.jetbrains.annotations.NotNull;

import java.nio.ByteBuffer;

/**
 * Created by dsmith on 11/11/16.
 */
public class MarshallableObj implements Marshallable {
    private final StringBuilder builder = new StringBuilder();

    public void clear() {
        builder.setLength(0);
    }

    public void append(CharSequence cs) {
        builder.append(cs);
    }

    @Override
    public void readMarshallable(WireIn wire) throws IORuntimeException {
        builder.setLength(0);
        Preconditions.checkState(wire.getValueIn().textTo(builder) != null);
    }

    @Override
    public void writeMarshallable(WireOut wire) {
        wire.getValueOut().text(builder);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        MarshallableObj that = (MarshallableObj) o;

        return builder.toString().equals(that.builder.toString());
    }

    @Override
    public int hashCode() {
        return builder.toString().hashCode();
    }

    @Override
    public String toString() {
        return builder.toString();
    }

    public static class Outer implements Marshallable {
        private final MarshallableObj obj = new MarshallableObj();

        public MarshallableObj getObj() {
            return obj;
        }

        @Override
        public void readMarshallable(@NotNull WireIn wire) throws IORuntimeException {
            wire.read(() -> "obj").marshallable(obj);
        }

        @Override
        public void writeMarshallable(@NotNull WireOut wire) {
            wire.write(() -> "obj").marshallable(obj);
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            Outer outer = (Outer) o;

            return obj.equals(outer.obj);

        }

        @Override
        public int hashCode() {
            return obj.hashCode();
        }
    }

    public static void main(String[] args) {
        final WireType wireType = WireType.TEXT;
        final String exampleString = "{";

        final Outer obj1 = new Outer();
        final Outer obj2 = new Outer();

        obj1.getObj().append(exampleString);

        final Bytes<ByteBuffer> bytes = Bytes.elasticByteBuffer();
        obj1.writeMarshallable(wireType.apply(bytes));

        final String output = bytes.toString();
        System.out.println("output: [" + output + "]");

        obj2.readMarshallable(wireType.apply(Bytes.from(output)));

        System.out.println("obj1.equals(obj2): " + obj1.equals(obj2));
    }
}
```

Changing to `exampleString="}"` executes with the `equals` being false.

Both examples successfully execute w/ `WireType.BINARY`.

Given this, and https://github.com/OpenHFT/Chronicle-Wire/issues/37, it seems like TEXT isn't really ready for prime time?

**Comment by peter-lawrey** [comment]
I have converted this into WireBug38Test. Can you review?

**Comment by devinrsmith** [comment]
LGTM. 

On a larger not, not sure how your testing infra is setup, but it might be worth it to have a bunch of marshallable objects that go through a suite of tests given the different `WireType`s.

**Comment by peter-lawrey** [comment]
Have a look at Json222Test where we added 222 tests used to validate JSON
parsers. The same data values are used for YAML. This showed up a number of
bugs which appear to be related to the ones you reported.

**Comment by hft-team-city** [comment]
Released in [Chronicle-Wire-2.20.101](https://github.com/OpenHFT/Chronicle-Wire/releases/chronicle-wire-2.20.101), [BOM-2.20.134](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.20.134)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Wire-2.22ea11](https://github.com/OpenHFT/Chronicle-Wire/releases/chronicle-wire-2.22ea11), [BOM-2.22ea69](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.22ea69)
